core/src/main/scala/reconciler/Reconciler.scala

```diff
@@ -119,6 +119,88 @@ abstract class Reconciler[S <: CustomResource[_, _], T <: ObjectResource] {
   }
   // scalastyle:on
 
+  case class CacheEntry(s: Option[Source], ts: List[Target])
+  case class CacheEntryKey(name: String, namespace: String, kind: String)
+  case class Cache(events: List[Event], cache: Map[CacheEntryKey, CacheEntry])
+
+  def watchSourceRaw(implicit context: RequestContext,
+                     sourceFormat: Format[S],
+                     sourceListFormat: Format[ListResource[S]],
+                     sourceResourceDefinition: ResourceDefinition[S],
+                     sourceListResourceDefinition: ResourceDefinition[ListResource[S]],
+                     ec: ExecutionContext): Future[akka.stream.scaladsl.Source[WatchEvent[S], NotUsed]] = {
+    context.list[ListResource[S]].map { l =>
+      val initialSource = Source(l.items.map(l => WatchEvent(EventType.MODIFIED, l)))
+      val watchedSource = context.watchAllContinuously[S](Some(l.resourceVersion))
+
+      initialSource.concat(watchedSource)
+    }
+  }
+
+  def watchTargetRaw(implicit context: RequestContext,
+                     targetFormat: Format[T],
+                     targetListFormat: Format[ListResource[T]],
+                     targetResourceDefinition: ResourceDefinition[T],
+                     targetListResourceDefinition: ResourceDefinition[ListResource[T]],
+                     ec: ExecutionContext): Future[akka.stream.scaladsl.Source[WatchEvent[T], NotUsed]] = {
+    context.list[ListResource[T]].map { l =>
+      val initialSource = Source(l.items.map(l => WatchEvent(EventType.MODIFIED, l)))
+      val watchedSource = context.watchAllContinuously[T](Some(l.resourceVersion))
+
+      initialSource.concat(watchedSource)
```

Comment:
Rather than creating the initialSource and passing the events through the stream it is probably better to build an initial state that can be pasted to the scan and the initial state of the system. In that way, when you start to reconcile the initial state of the system you will have both the source and target state. This will ensure that you don't make unnecessary requests to K8s.

Reply:
Thanks for pointing out that initial state can be easily built, I've made the necessary changes.